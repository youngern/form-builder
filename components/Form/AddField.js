import React from 'react';
import {
  SafeAreaView,
  StyleSheet,
  View,
  Text,
} from 'react-native';
import _ from 'lodash';
import { Form, Field } from 'react-final-form';
import { Colors } from 'react-native/Libraries/NewAppScreen';
import { Header, Icon } from 'react-native-elements';
import Input from './Input';
import Error from './Error';
import FormButton from './FormButton';

const required = (value) => value ? undefined : 'required';
const App = (props) => {
  const { onAdd } = props;

  return (
    <>
      <Header
        containerStyle={{ borderBottomWidth: undefined }}
        backgroundColor="#fff"
        leftComponent={(
          <Icon
            name='clear'
            type='material-icons'
            color='#4d80e4'
            onPress={() => { setModalVisible(false); }}
          />
        )}
      />
      <Form onSubmit={onAdd}>
        {({ handleSubmit }) => (
          <SafeAreaView style={{ flex: 1 }}>
            <View>
              <Text style={styles.fieldHeader}>Add Field</Text>
              <Field
                subscription={{ value: true, active: true, touched: true, error: true }}
                name="name"
                placeholder="Name"
                validate={required}
              >
                {({ input, meta, ...rest }) => {
                  const error = (meta.touched && meta.error) || undefined;

                  return (
                    <Input
                      error={error}
                      {...input}
                      {...rest}
                    />
                  )
                }}
              </Field>
              <Error name="name" />
              <Field
                subscription={{ value: true, active: true, touched: true, error: true }}
                name="placeholder"
                placeholder="Placeholder"
              >
                {({ input, meta, ...rest }) => {
                  const error = (meta.touched && meta.error) || undefined;

                  return (
                    <Input
                      error={error}
                      {...input}
                      {...rest}
                    />
                  )
                }}
              </Field>
              <Error name="name" />
              <Field
                subscription={{ value: true, active: true, touched: true, error: true }}
                name="initialValue"
                placeholder="Initial Value"
              >
                {({ input, meta, ...rest }) => {
                  const error = (meta.touched && meta.error) || undefined;

                  return (
                    <Input
                      error={error}
                      {...input}
                      {...rest}
                    />
                  )
                }}
              </Field>
              <Error name="initialValue" />
            </View>
            <FormButton
              label="Add Field"
              onPress={handleSubmit}
              style={[styles.submit, { flex: 0 }]}
              labelStyle={{ fontSize: 20 }}
            />
          </SafeAreaView>
        )}
      </Form>
    </>
  );
};

const baseColor = '#f3fcf9';

const styles = StyleSheet.create({
  container: {
    backgroundColor: baseColor,
    flex: 1,
  },

  footer: {
    backgroundColor: baseColor,
    flexDirection: 'row',
    justifyContent: 'space-around',
    paddingBottom: 25,
  },

  formHeader: {
    borderWidth: 0,
    backgroundColor: baseColor,
    fontSize: 30,
    color: '#4d80e4',
    padding: 0,
    marginLeft: 0,
  },

  fieldHeader: {
    borderWidth: 0,
    fontSize: 30,
    color: '#4d80e4',
    padding: 0,
    marginLeft: 10,
    marginBottom: 10,
  },

  spy: {
    backgroundColor: Colors.white,
    padding: 10,
    marginHorizontal: 10,
    borderColor: 'lightgray',
    borderRadius: 5,
    borderWidth: 1,
  },

  submit: {
    backgroundColor: '#4d80e4',
    borderColor: '#4d80e4'
  },

  reset: {
    backgroundColor: '#fbfefd',
    borderColor: '#46b3e6',
  },
  
  active: {
    borderColor: '#46b3e6',
  },
});

export default App;
