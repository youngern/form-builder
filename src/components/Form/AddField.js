import React from 'react';
import { SafeAreaView, StyleSheet, View, Text, Switch } from 'react-native';
import _ from 'lodash';
import { Form, Field } from 'react-final-form';

import Config from '~/src/services/config';
import Input from './Input';
import Error from './Error';
import Label from './Label';
import FormButton from './FormButton';

const { Colors } = Config;
const SwitchInput = React.forwardRef((props, ref) => {
  const { onChange, ...rest } = props;
  return (
    <View
      style={{
        paddingRight: 5,
        paddingVertical: 10,
        marginVertical: 5,
        alignItems: 'flex-start',
      }}>
      <Switch
        ref={ref}
        trackColor={{ false: '#edffea', true: '#75daad' }}
        thumbColor="#fff"
        ios_backgroundColor="#ed6663"
        onValueChange={onChange}
        {...rest}
      />
    </View>
  );
});

const required = (value) => (value ? undefined : 'required');
const App = (props) => {
  const { values = undefined, onAdd } = props;
  return (
    <Form onSubmit={onAdd} initialValues={values}>
      {({ handleSubmit }) => (
        <SafeAreaView style={{ flex: 1 }}>
          <View>
            <Text style={styles.fieldHeader}>Add Field</Text>
            <View style={styles.section}>
              <Label value="Label" />
              <Field
                subscription={{
                  value: true,
                  active: true,
                  touched: true,
                  error: true,
                }}
                name="label"
                placeholder="Label"
                validate={required}>
                {({ input, meta, ...rest }) => {
                  const error = (meta.touched && meta.error) || undefined;

                  return <Input error={error} {...input} {...rest} />;
                }}
              </Field>
              <Error name="label" />
            </View>
            <View style={styles.section}>
              <Label value="Description" />
              <Field
                subscription={{
                  value: true,
                  active: true,
                  touched: true,
                  error: true,
                }}
                name="description"
                placeholder="Description">
                {({ input, meta, ...rest }) => {
                  const error = (meta.touched && meta.error) || undefined;

                  return <Input error={error} {...input} {...rest} />;
                }}
              </Field>
            </View>
            <Error name="description" />
            <View style={styles.section}>
              <Label value="Placeholder" />
              <Field
                subscription={{
                  value: true,
                  active: true,
                  touched: true,
                  error: true,
                }}
                name="placeholder"
                placeholder="Placeholder">
                {({ input, meta, ...rest }) => {
                  const error = (meta.touched && meta.error) || undefined;

                  return <Input error={error} {...input} {...rest} />;
                }}
              </Field>
              <Error name="placeholder" />
            </View>
            <View style={styles.section}>
              <Label value="Initial Value" />
              <Field
                subscription={{
                  value: true,
                  active: true,
                  touched: true,
                  error: true,
                }}
                name="initialValue"
                placeholder="Initial Value">
                {({ input, meta, ...rest }) => {
                  const error = (meta.touched && meta.error) || undefined;

                  return <Input error={error} {...input} {...rest} />;
                }}
              </Field>
              <Error name="initialValue" />
            </View>
            <View style={styles.section}>
              <Label value="Required" />
              <Field
                subscription={{
                  value: true,
                  active: true,
                  touched: true,
                  error: true,
                }}
                name="required"
                placeholder="Required"
                initialValue={false}>
                {({ input, meta, ...rest }) => {
                  const error = (meta.touched && meta.error) || undefined;

                  return <SwitchInput error={error} {...input} {...rest} />;
                }}
              </Field>
              <Error name="required" />
            </View>
          </View>
          <FormButton
            label="Add Field"
            onPress={handleSubmit}
            style={[styles.submit, { flex: 0 }]}
            labelStyle={{ fontSize: 20 }}
          />
        </SafeAreaView>
      )}
    </Form>
  );
};

const styles = StyleSheet.create({
  container: {
    backgroundColor: Colors.background,
    flex: 1,
  },

  footer: {
    backgroundColor: Colors.background,
    flexDirection: 'row',
    justifyContent: 'space-around',
    paddingBottom: 25,
  },

  formHeader: {
    borderWidth: 0,
    backgroundColor: Colors.background,
    fontSize: 30,
    color: '#4d80e4',
    padding: 0,
    marginLeft: 0,
  },

  fieldHeader: {
    borderWidth: 0,
    fontSize: 30,
    color: '#4d80e4',
    padding: 0,
    marginLeft: 10,
    marginBottom: 10,
  },

  spy: {
    backgroundColor: Colors.white,
    padding: 10,
    marginHorizontal: 10,
    borderColor: 'lightgray',
    borderRadius: 5,
    borderWidth: 1,
  },

  submit: {
    backgroundColor: '#4d80e4',
    borderColor: '#4d80e4',
  },

  reset: {
    backgroundColor: '#fbfefd',
    borderColor: '#46b3e6',
  },

  active: {
    borderColor: '#46b3e6',
  },

  section: {
    marginHorizontal: 10,
  },
});

export default App;
