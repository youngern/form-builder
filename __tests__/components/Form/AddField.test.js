import React from 'react';
import AddField from 'final-form-react-native/src/components/Form/AddField';
import renderer from 'react-test-renderer';

it('renders correctly with initial values', () => {
  const tree = renderer
    .create(
      <AddField
        values={{ initialValue: 'nada', fieldIndex: 0 }}
        onAdd={() => {}}
      />,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders correctly without initial values', () => {
  const tree = renderer
    .create(<AddField values={undefined} onAdd={() => {}} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
