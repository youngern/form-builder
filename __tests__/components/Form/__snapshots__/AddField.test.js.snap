// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with initial values 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View>
    <Text
      style={
        Object {
          "borderWidth": 0,
          "color": "#4d80e4",
          "fontSize": 30,
          "marginBottom": 10,
          "marginLeft": 10,
          "padding": 0,
        }
      }
    >
      Add Field
    </Text>
    <View
      style={
        Object {
          "marginHorizontal": 10,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#4d80e4",
            "fontFamily": "AppleSDGothicNeo-Bold",
            "fontSize": 14,
          }
        }
      >
        label
      </Text>
      <TextInput
        allowFontScaling={true}
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Lame"
        rejectResponderTermination={true}
        style={
          Array [
            Object {
              "backgroundColor": "#fbfefd",
              "borderColor": "lightgray",
              "borderRadius": 5,
              "borderWidth": 1,
              "color": "#204051",
              "fontFamily": "AppleSDGothicNeo-Bold",
              "fontSize": 20,
              "paddingLeft": 10,
              "paddingRight": 5,
              "paddingVertical": 10,
            },
            undefined,
            undefined,
          ]
        }
        underlineColorAndroid="transparent"
        value=""
      />
      <View
        style={
          Object {
            "height": 14,
          }
        }
      />
    </View>
    <View
      style={
        Object {
          "marginHorizontal": 10,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#4d80e4",
            "fontFamily": "AppleSDGothicNeo-Bold",
            "fontSize": 14,
          }
        }
      >
        description
      </Text>
      <TextInput
        allowFontScaling={true}
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Description"
        rejectResponderTermination={true}
        style={
          Array [
            Object {
              "backgroundColor": "#fbfefd",
              "borderColor": "lightgray",
              "borderRadius": 5,
              "borderWidth": 1,
              "color": "#204051",
              "fontFamily": "AppleSDGothicNeo-Bold",
              "fontSize": 20,
              "paddingLeft": 10,
              "paddingRight": 5,
              "paddingVertical": 10,
            },
            undefined,
            undefined,
          ]
        }
        underlineColorAndroid="transparent"
        value=""
      />
    </View>
    <View
      style={
        Object {
          "height": 14,
        }
      }
    />
    <View
      style={
        Object {
          "marginHorizontal": 10,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#4d80e4",
            "fontFamily": "AppleSDGothicNeo-Bold",
            "fontSize": 14,
          }
        }
      >
        placeholder
      </Text>
      <TextInput
        allowFontScaling={true}
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Placeholder"
        rejectResponderTermination={true}
        style={
          Array [
            Object {
              "backgroundColor": "#fbfefd",
              "borderColor": "lightgray",
              "borderRadius": 5,
              "borderWidth": 1,
              "color": "#204051",
              "fontFamily": "AppleSDGothicNeo-Bold",
              "fontSize": 20,
              "paddingLeft": 10,
              "paddingRight": 5,
              "paddingVertical": 10,
            },
            undefined,
            undefined,
          ]
        }
        underlineColorAndroid="transparent"
        value=""
      />
      <View
        style={
          Object {
            "height": 14,
          }
        }
      />
    </View>
    <View
      style={
        Object {
          "marginHorizontal": 10,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#4d80e4",
            "fontFamily": "AppleSDGothicNeo-Bold",
            "fontSize": 14,
          }
        }
      >
        initialValue
      </Text>
      <TextInput
        allowFontScaling={true}
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Initial Value"
        rejectResponderTermination={true}
        style={
          Array [
            Object {
              "backgroundColor": "#fbfefd",
              "borderColor": "lightgray",
              "borderRadius": 5,
              "borderWidth": 1,
              "color": "#204051",
              "fontFamily": "AppleSDGothicNeo-Bold",
              "fontSize": 20,
              "paddingLeft": 10,
              "paddingRight": 5,
              "paddingVertical": 10,
            },
            undefined,
            undefined,
          ]
        }
        underlineColorAndroid="transparent"
        value="nada"
      />
      <View
        style={
          Object {
            "height": 14,
          }
        }
      />
    </View>
    <View
      style={
        Object {
          "marginHorizontal": 10,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#4d80e4",
            "fontFamily": "AppleSDGothicNeo-Bold",
            "fontSize": 14,
          }
        }
      >
        required
      </Text>
      <View
        style={
          Object {
            "alignItems": "flex-start",
            "marginVertical": 5,
            "paddingRight": 5,
            "paddingVertical": 10,
          }
        }
      >
        <RCTSwitch
          accessibilityRole="switch"
          name="required"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          onTintColor="#75daad"
          placeholder="Required"
          style={
            Array [
              Object {
                "height": 31,
                "width": 51,
              },
              Object {
                "backgroundColor": "#ed6663",
                "borderRadius": 16,
              },
            ]
          }
          thumbTintColor="#fff"
          tintColor="#edffea"
          value={false}
        />
      </View>
      <View
        style={
          Object {
            "height": 14,
          }
        }
      />
    </View>
  </View>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#4d80e4",
        "borderColor": "#4d80e4",
        "borderRadius": 5,
        "borderWidth": 1,
        "flex": 0,
        "margin": 10,
        "opacity": 1,
        "padding": 10,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#FFF",
            "fontWeight": "bold",
          },
          Object {
            "fontSize": 20,
          },
        ]
      }
    >
      Add Field
    </Text>
  </View>
</RCTSafeAreaView>
`;

exports[`renders correctly without initial values 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View>
    <Text
      style={
        Object {
          "borderWidth": 0,
          "color": "#4d80e4",
          "fontSize": 30,
          "marginBottom": 10,
          "marginLeft": 10,
          "padding": 0,
        }
      }
    >
      Add Field
    </Text>
    <View
      style={
        Object {
          "marginHorizontal": 10,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#4d80e4",
            "fontFamily": "AppleSDGothicNeo-Bold",
            "fontSize": 14,
          }
        }
      >
        label
      </Text>
      <TextInput
        allowFontScaling={true}
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Lame"
        rejectResponderTermination={true}
        style={
          Array [
            Object {
              "backgroundColor": "#fbfefd",
              "borderColor": "lightgray",
              "borderRadius": 5,
              "borderWidth": 1,
              "color": "#204051",
              "fontFamily": "AppleSDGothicNeo-Bold",
              "fontSize": 20,
              "paddingLeft": 10,
              "paddingRight": 5,
              "paddingVertical": 10,
            },
            undefined,
            undefined,
          ]
        }
        underlineColorAndroid="transparent"
        value=""
      />
      <View
        style={
          Object {
            "height": 14,
          }
        }
      />
    </View>
    <View
      style={
        Object {
          "marginHorizontal": 10,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#4d80e4",
            "fontFamily": "AppleSDGothicNeo-Bold",
            "fontSize": 14,
          }
        }
      >
        description
      </Text>
      <TextInput
        allowFontScaling={true}
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Description"
        rejectResponderTermination={true}
        style={
          Array [
            Object {
              "backgroundColor": "#fbfefd",
              "borderColor": "lightgray",
              "borderRadius": 5,
              "borderWidth": 1,
              "color": "#204051",
              "fontFamily": "AppleSDGothicNeo-Bold",
              "fontSize": 20,
              "paddingLeft": 10,
              "paddingRight": 5,
              "paddingVertical": 10,
            },
            undefined,
            undefined,
          ]
        }
        underlineColorAndroid="transparent"
        value=""
      />
    </View>
    <View
      style={
        Object {
          "height": 14,
        }
      }
    />
    <View
      style={
        Object {
          "marginHorizontal": 10,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#4d80e4",
            "fontFamily": "AppleSDGothicNeo-Bold",
            "fontSize": 14,
          }
        }
      >
        placeholder
      </Text>
      <TextInput
        allowFontScaling={true}
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Placeholder"
        rejectResponderTermination={true}
        style={
          Array [
            Object {
              "backgroundColor": "#fbfefd",
              "borderColor": "lightgray",
              "borderRadius": 5,
              "borderWidth": 1,
              "color": "#204051",
              "fontFamily": "AppleSDGothicNeo-Bold",
              "fontSize": 20,
              "paddingLeft": 10,
              "paddingRight": 5,
              "paddingVertical": 10,
            },
            undefined,
            undefined,
          ]
        }
        underlineColorAndroid="transparent"
        value=""
      />
      <View
        style={
          Object {
            "height": 14,
          }
        }
      />
    </View>
    <View
      style={
        Object {
          "marginHorizontal": 10,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#4d80e4",
            "fontFamily": "AppleSDGothicNeo-Bold",
            "fontSize": 14,
          }
        }
      >
        initialValue
      </Text>
      <TextInput
        allowFontScaling={true}
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Initial Value"
        rejectResponderTermination={true}
        style={
          Array [
            Object {
              "backgroundColor": "#fbfefd",
              "borderColor": "lightgray",
              "borderRadius": 5,
              "borderWidth": 1,
              "color": "#204051",
              "fontFamily": "AppleSDGothicNeo-Bold",
              "fontSize": 20,
              "paddingLeft": 10,
              "paddingRight": 5,
              "paddingVertical": 10,
            },
            undefined,
            undefined,
          ]
        }
        underlineColorAndroid="transparent"
        value=""
      />
      <View
        style={
          Object {
            "height": 14,
          }
        }
      />
    </View>
    <View
      style={
        Object {
          "marginHorizontal": 10,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#4d80e4",
            "fontFamily": "AppleSDGothicNeo-Bold",
            "fontSize": 14,
          }
        }
      >
        required
      </Text>
      <View
        style={
          Object {
            "alignItems": "flex-start",
            "marginVertical": 5,
            "paddingRight": 5,
            "paddingVertical": 10,
          }
        }
      >
        <RCTSwitch
          accessibilityRole="switch"
          name="required"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          onTintColor="#75daad"
          placeholder="Required"
          style={
            Array [
              Object {
                "height": 31,
                "width": 51,
              },
              Object {
                "backgroundColor": "#ed6663",
                "borderRadius": 16,
              },
            ]
          }
          thumbTintColor="#fff"
          tintColor="#edffea"
          value={false}
        />
      </View>
      <View
        style={
          Object {
            "height": 14,
          }
        }
      />
    </View>
  </View>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#4d80e4",
        "borderColor": "#4d80e4",
        "borderRadius": 5,
        "borderWidth": 1,
        "flex": 0,
        "margin": 10,
        "opacity": 1,
        "padding": 10,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#FFF",
            "fontWeight": "bold",
          },
          Object {
            "fontSize": 20,
          },
        ]
      }
    >
      Add Field
    </Text>
  </View>
</RCTSafeAreaView>
`;
